src/gluonts/mx/distribution/ISQF.py

```diff
@@ -0,0 +1,864 @@
+        # d: d_s, d_plus: d_{s+1}, delta_d: d_{s+1}-d_s
+        # for s=0,...,num_pieces-1
+        # shape=(*batch_shape, num_qk-1, num_pieces)
+        self.p, self.delta_p = ISQF.parametrize_spline(
+            F, self.heights, self.q, self.q_plus, self.num_pieces, self.tol
+        )
+        self.d, self.delta_d = ISQF.parametrize_spline(
+            F,
+            self.knots,
+            self.alpha,
+            self.alpha_plus,
+            self.num_pieces,
+            self.tol,
+        )
+
+        if self.num_pieces > 1:
+            self.d_plus = F.concat(
+                F.slice_axis(self.d, axis=-1, begin=1, end=None),
+                F.expand_dims(self.alpha_plus, axis=-1),
+                dim=-1,
+            )
+        else:
+            self.d_plus = F.expand_dims(self.alpha_plus, axis=-1)
+
+        # Get tails parameters
+        self.a_l, self.b_l = ISQF.parametrize_tail(
+            F, self.beta_l, self.alpha_l, self.q_l
+        )
+        self.a_r, self.b_r = ISQF.parametrize_tail(
+            F, -self.beta_r, 1 - self.alpha_r, self.q_r
+        )
+
+    @property
+    def F(self):
+        return getF(self.beta_l)
+
+    @property
+    def args(self) -> List:
+        return [
+            self.knots,
+            self.heights,
+            self.q_proj,
+            self.beta_l,
+            self.beta_r,
+        ]
+
+    @staticmethod
+    def parametrize_qk(
+        F, q_proj: Tensor
+    ) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
+        # Refer to the descriptions in init for details of parameters
+        q = F.slice_axis(q_proj, axis=-1, begin=0, end=-1)
+        q_plus = F.slice_axis(q_proj, axis=-1, begin=1, end=None)
+        q_l = F.slice_axis(q_proj, axis=-1, begin=0, end=1).squeeze(axis=-1)
+        q_r = F.slice_axis(q_proj, axis=-1, begin=-1, end=None).squeeze(
+            axis=-1
+        )
+
+        return q, q_plus, q_l, q_r
+
+    @staticmethod
+    def parametrize_spline(
+        F,
+        knots: Tensor,
+        alpha: Tensor,
+        alpha_plus: Tensor,
+        num_pieces: int,
+        tol: float = 1e-4,
+    ) -> Tuple[Tensor, Tensor]:
+        # Refer to the descriptions in init for details of parameters
+
+        # The spacing between spline knots is parametrized by softmax function (in [0,1] and sum to 1)
+        # We add tol to prevent overflow when we compute 1/spacing in spline CRPS
+        # After adding tol, it is normalized by (1 + num_pieces * tol) to keep the sum-to-1 property
+        delta_x = (F.softmax(knots) + tol) / (1 + num_pieces * tol)
+        x = cumsum(F, delta_x, exclusive=True)
+
+        alpha = F.expand_dims(alpha, axis=-1)
+        alpha_plus = F.expand_dims(alpha_plus, axis=-1)
+
+        d = F.broadcast_add(F.broadcast_mul(x, (alpha_plus - alpha)), alpha)
+        delta_d = F.broadcast_mul(delta_x, (alpha_plus - alpha))
+
+        return d, delta_d
+
+    @staticmethod
+    def parametrize_tail(
+        F, beta: Tensor, alpha: Tensor, q: Tensor
```

Comment:
Add comment description on parameters, and same for below methods.

Reply:
I put their descriptions in init. I will move them to the functions.